AI Chat Tool definition and EngineConsumer inclusion

petemill opened the pull request:
Depends on:
- https://github.com/brave/brave-core/pull/29643

A Tool is a feature the LLM can "use" by sending a tool request with some input parameters specified. This PR defines the Tool base class and determines where to send the array of Tools in use for a particular engine request: `EngineConsumer::GenerateAssistantResponse`. When the conversation is deciding which tools to use, it will consult `Tool::IsSupportedByModel()`, `Tool::IsAgentTool()`, `Tool::IsContentAssociationRequired()` as well as knowledge about a static list of Tools.

When an engine responds with a tool use request, that request will be added to the events of a conversation history entry (via `ToolUseEvent` - not implemented in this PR). The Tool class in this PR will be used to provide the result of that tool use request, via `Tool::UseTool()`. We will then provide a response to the tool in the `ToolUseEvent`'s output field (not implemented in this PR). The LLM will be called again providing an additional role=Tool conversation history entry.

In order to keep other submissions' size as small as possible, and to submit in parallel, this PR only contains the base definition of Tool and where Tool gets passed. It does not contain the tool uses, the related `ToolUseEvent` tool use and response conversation event, any Engine implementation which passes a list of tools to an API, or any UI representation of Tools or tool uses.

Subsequent submissions will be:
- Conversation API tool handling
- OAI Chat API tool handling
- ConversationHandler getting an agentic loop to perform the usage of the Tools and send the responses back to the engine
- UI representation for tool uses
- Tool implementations

<!-- Add brave-browser issue below that this PR will resolve -->
Resolves:
- Resolves https://github.com/brave/brave-browser/issues/46977

<!-- CI-related labels that can be applied to this PR:
* CI/run-audit-deps (1) - check for known npm/cargo vulnerabilities (audit_deps)
* CI/run-network-audit (1) - run network-audit
* CI/run-upstream-tests - run Chromium unit and browser tests on Linux and Windows (otherwise only on Linux)
* CI/run-linux-arm64, CI/run-macos-x64, CI/run-windows-arm64, CI/run-windows-x86 - run builds that would otherwise be skipped
* CI/skip - do not run CI builds (except noplatform)
* CI/skip-linux-x64, CI/skip-android, CI/skip-macos-arm64, CI/skip-ios, CI/skip-windows-x64 - skip CI builds for specific platforms
* CI/skip-upstream-tests - do not run Chromium unit, or browser tests (otherwise only on Linux)
* CI/skip-all-linters - do not run presubmit and lint checks
* CI/storybook-url (1) - deploy storybook and provide a unique URL for each build

(1) applied automatically when some files are changed (see: https://github.com/brave/brave-core/blob/master/.github/labeler.yml)
-->

<!--
## Checklist:

- Review design docs
  [Browser design principles](https://chromium.googlesource.com/chromium/src/+/refs/heads/main/docs/chrome_browser_design_principles.md)
  [Style guide](https://chromium.googlesource.com/chromium/src/+/main/styleguide/c++/c++.md)
  [Core principles](https://www.chromium.org/developers/core-principles/)
- Ensure there are (tests)[https://www.chromium.org/developers/testing/]. Unit test as much as possible (including edge cases), but also include browser tests covering high level functionality.
- Ensure that there are comments explaining what classes/methods are/do. The "why" is often more important than the "what" in comments. Also update any relevant docs (moving docs from wiki to brave-core if necessary).
- Request security or other review (third-party libraries, rust code, etc...) if applicable [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) [other review](https://github.com/brave/reviews/issues/new/choose)
  Also see [adding third-party libraries](https://chromium.googlesource.com/chromium/src/+/refs/heads/main/docs/adding_to_third_party.md) for general guidelines on using third party code
- Maks sure there is a [ticket](https://github.com/brave/brave-browser/issues) for your issue
- Use Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- Write a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- Squash any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- Add appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- Run `git rebase master` (if needed)
-->

yrliou commented on components/ai_chat/core/browser/model_service.cc:
```diff
@@ -113,6 +113,7 @@ const std::vector<mojom::ModelPtr>& GetLeoModels() {
       model->key = "chat-automatic";
       model->display_name = "Automatic";
       model->vision_support = true;
+      model->supports_tools = false;
```
I think it's better to set this to true and have a small server side change to distribute requests to supported models if tools event present like we did for vision.

yrliou replied:
Ok, it probably needs more thoughts on server side when triaging requests @Steeeephen to decide to use tools or not then. Not in the scope of this PR, just that limiting tools to a explicitly model choice would somehow defeat the purpose of having an automatic model.

petemill replied:
Hopefully there'll be a time very soon when all models will support tools

Steeeephen replied:
Not 100% sure how it works on the browser/agentic browser side but in general I would think the less hardcoding we can do in the browser the better, dealing with things like this in the Leo backend lets us move quicker

yrliou replied:
@Steeeephen is it possible for server (the triaging model) to decide whether to use tools or not based on the request when client always pass tools?

yrliou commented on components/ai_chat/core/browser/engine/conversation_api_client.cc:
```diff
@@ -217,14 +218,15 @@ void ConversationAPIClient::ClearAllQueries() {
   auto callback = base::BindOnce(
       &ConversationAPIClient::PerformRequestWithCredentials,
-      weak_ptr_factory_.GetWeakPtr(), std::move(conversation),
+      weak_ptr_factory_.GetWeakPtr(), std::move(conversation), tools,
```
As mentioned in the DM, it isn't safe to bind the vector of pointers here.

petemill replied:
removed from this PR, we can discuss in a subsequent one

brave-builds commented:
A Storybook has been deployed to [preview UI for the latest push](https://build-artifacts.brave.com/brave-browser-build-pr/aichat-tools-definition/storybook.html)

petemill commented on components/ai_chat/core/browser/tools/tool.h:
```diff
@@ -0,0 +1,86 @@
+class Tool {
+ public:
+  using ToolResult = std::optional<std::vector<mojom::ContentBlockPtr>>;
+  using UseToolCallback = base::OnceCallback<void(ToolResult&& output)>;
```
Removed from this PR, we can discuss in a subsequent PR where this was seemingly necessary.

petemill commented on components/ai_chat/core/browser/engine/engine_consumer.h:
```diff
@@ -85,6 +89,8 @@ class EngineConsumer {
       const std::string& page_content,
       const ConversationHistory& conversation_history,
       const std::string& selected_language,
+      const Tools tools,
```
Changed to make sure the references to the tools are not held on in any callback at all

darkdh commented on components/ai_chat/core/browser/tools/tool_input_properties.cc:
```diff
@@ -0,0 +1,95 @@
+    property.Set("description", description);
+  }
+
+  if (enum_values.has_value() && enum_values->size() > 0) {
```
nit: `!enums_values->empty()`

darkdh commented on components/ai_chat/core/browser/tools/tool_input_properties.cc:
```diff
@@ -0,0 +1,95 @@
+
+  base::Value::Dict properties_dict;
+  for (auto& [key, value] : properties) {
+    properties_dict.Set(key, std::move(const_cast<base::Value::Dict&>(value)));
```
This should just call `Clone()`, const_cast and move it will eventually do just copy

petemill replied:
Great point - clone seems ok for performance here because we're discarding the input and since it's an initializer_list it's unlikely to have been copied.

darkdh commented on components/ai_chat/core/browser/ai_chat_service_unittest.cc:
```diff
@@ -446,11 +447,13 @@ TEST_P(AIChatServiceUnitTest,
+  EXPECT_CALL(*engine, GenerateAssistantResponse)
       .WillOnce(
-          [&resolve](bool send_page_contents, const std::string& page_contents,
+          [&resolve](const bool& is_video, const std::string& page_contents,
```
use just bool, bool is one byte and reference is bigger than that

petemill replied:
Seems sensible, but I was just making it consistent with the actual class implementation which is `const bool& is_video` so seems like we should change it in `EngineConsumer`. OK to keep here for consistency?

darkdh replied:
yeah, I just realized that is out of scope of this PR. Maybe you can fix it in other PRs.

darkdh commented on components/ai_chat/core/browser/engine/engine_consumer.h:
```diff
@@ -85,6 +87,8 @@ class EngineConsumer {
       const std::string& page_content,
       const ConversationHistory& conversation_history,
       const std::string& selected_language,
+      const std::vector<const Tool*> tools,
```
This should be `const std::vector<base::WeakPtr<Tool>>&` so we can guarantee the pointer is valid when we reference it. (It will be set to to null when the object is destroyed)

petemill commented:
@darkdh I addressed your feedback - thanks for the review https://github.com/brave/brave-core/compare/a62039285cf1a39eac76735f543a04d6a32468da..078b7eff6d10c2afb7c45b4596f61ab8032e6bf1

darkdh commented on components/ai_chat/core/browser/tools/tool.cc:
```diff
@@ -0,0 +1,52 @@
+
+void Tool::UseTool(const std::string& input_json,
+                   Tool::UseToolCallback callback) {
+  CHECK(false) << "UseTool called but not implemented";
```
nit: `NOTIMPLEMENTED()`

petemill replied:
I considered `NOTIMPLEMENTED` but I didn't think it was appropriate given the following: it's a base class and the expectation is that the derived class _may_ implement it, and we should never be running the base implementation.
I _would_ consider deleting the base class function but not _all_ consuming classes will implement their own `UseTool`, and it's optional as long as the tool running loop (introduced in another PR) knows about that.

For example, there will be some tools which do not provide a result themselves. For example when the tool requires a user response, or when the tool is a request to retrieve page content. In these cases, they are built-in `ConversationHandler` tools, and the conversation handler or the conversation UI will create the tool result based on user actions. The `CHECK` is therefore to make sure we're not calling `UseTool` on a tool that hasn't implemented `UseTool`. We want to either be implementing `UseTool` or creating the output of the tool elsewhere.

So, it is a default implementation, but running the default implementation should error. We could change this to be a default implementation that does nothing, which might be less confusing.

I have removed this function given it won't be possible to review what's best here without the rest of the tool loop PR, which I'm submitting now 😄 . Thanks
